packages/shared/src/lib/group_utils.ts

```diff
@@ -0,0 +1,71 @@
+// Perhaps we can remove some of these functions if it becomes cumbersome to
+// implement these functions every time.  For example, `forEach` can be replaced
+// by map if we ignore the output
+interface Fillable<K, V> {
+  at(index: K): V | undefined;
```

Comment:
What does undefined mean here? Out of bounds or no value?

Reply:
In the case of `Grid` (the only class that implements `Fillable` at the moment, `at` can return `undefined` for a number of reasons:

```
  const g1 = new Grid(2, 2);
  // undefined, empty slot
  g1.at({ x: 1, y: 1 });

  const g2 = Grid.from2DArray([
    [1, 2],
    [3, 4],
  ]);
  // undefined, out-of-bounds
  g2.at({ x: 0, y: 5 });

  const g3 = Grid.from2DArray([
    [undefined, undefined],
    [undefined, undefined],
  ]);
  // undefined, T is undefined
  g3.at({ x: 0, y: 0 });
```

FWIW, this matches the behavior for native `Array`s (but I'm not opposed to changing it if it makes the developer experience better)

```
const a1 = new Array(5);
a1[0];  // undefined

const a2 = [1, 2, 3];
a2[5];  // undefined

const a3 = [undefined, undefined, undefined];
a3[0];  // undefined
```

Reply:
Thanks!